### PR TITLE
[test] sgen tests should behave nicer on slower machines.

### PR DESCRIPTION
**Pull request by lambdageek**
*  `sgen-domain-unload-2.cs` stop before timeout.
  *  `sgen-new-threads-dont-join-stw.cs` print progress dots.

### CODE REVIEW

File: mono/tests/sgen-new-threads-dont-join-stw.cs [hunk 1]
```diff
@@ -82,6 +82,7 @@ static void Timer_Elapsed(object sender, EventArgs e)
             {
                 while (count < num_threads)
                 {
+		    Console.Write(".");
```
Comment (akoeplinger) [inline mono/tests/sgen-new-threads-dont-join-stw.cs]: mixed tabs/spaces

File: mono/tests/test-driver [hunk 2]
```diff
@@ -29,9 +29,10 @@ if ($test =~ /.*\|.*/) {
 	#This is a silly workaround, but all tests that use extra parameters need a larger timeout.
 	$timeout_in_minutes = 5;
 }
+$ENV{'TEST_DRIVER_TIMEOUT_SEC'} = $timeout_in_minutes * 60;
```
Comment (akoeplinger) [inline mono/tests/test-driver]: We'll need the same change for the managed test runner (which is what Jenkins uses).

File: mono/tests/sgen-domain-unload-2.cs [hunk 3]
```diff
@@ -37,7 +40,32 @@ static void BackgroundNoise ()
...
+				break;
+			}
 		}
 		Console.WriteLine ("\ndone");
 	}
-}
\ No newline at end of file
+
+	// 5s slack
+	private static readonly TimeSpan slack = TimeSpan.FromMilliseconds(5000);
```
Comment (akoeplinger) [inline mono/tests/sgen-domain-unload-2.cs]: are we sure 5s is enough time even on slow systems?

File: mono/tests/sgen-domain-unload-2.cs [hunk 4]
```diff
@@ -9,6 +9,8 @@
 stop-the-world before walking the heap.
 */
 class Driver {
+	private static readonly string ENV_TIMEOUT = "TEST_DRIVER_TIMEOUT_SEC";
+	private static readonly TimeSpan testDuration = getTestDuration();
```
Comment (akoeplinger) [inline mono/tests/sgen-domain-unload-2.cs]: I'd just make this a property

File: mono/tests/sgen-domain-unload-2.cs [hunk 5]
```diff
@@ -24,6 +26,7 @@ static void BackgroundNoise ()
 	}
 
 	static void Main () {
+		var startTime = DateTime.Now;
```
Comment (akoeplinger) [inline mono/tests/sgen-domain-unload-2.cs]: use `DateTime.UtcNow` as Now is affected by DST et al.

File: mono/tests/sgen-domain-unload-2.cs [hunk 6]
```diff
@@ -37,7 +40,32 @@ static void BackgroundNoise ()
...
+		if (testDuration == TimeSpan.Zero)
+			return true;
+		var t = DateTime.Now;
+		return (start + testDuration + slack < t);
+	}
+
+	private static TimeSpan getTestDuration() {
+		var e = System.Environment.GetEnvironmentVariable(ENV_TIMEOUT);
+		double d;
+		if (e != null && Double.TryParse(e, out d)) {
```
Comment (akoeplinger) [inline mono/tests/sgen-domain-unload-2.cs]: the null check isn't needed, TryParse handles that already.

File: mono/tests/test-driver [hunk 7]
```diff
@@ -29,9 +29,10 @@ if ($test =~ /.*\|.*/) {
...
 
 $| = 0;
-print "Testing $test... ";
+print "Testing $test (TEST_DRIVER_TIMEOUT_SEC=$ENV{'TEST_DRIVER_TIMEOUT_SEC'})... ";
```
Comment (akoeplinger) [inline mono/tests/test-driver]: I'm not sure we really need to output the env var here, it'll just spam the output.

File: mono/tests/test-driver (hunk 2 above)
Comment (lambdageek) [inline mono/tests/test-driver]: So interestingly, the sgen tests always run with the Perl runner.  But I went ahead and added the env setting to the maanged runner too if it turns out useful in the future.

File: mono/tests/sgen-domain-unload-2.cs (hunk 3 above)
Comment (lambdageek) [inline mono/tests/sgen-domain-unload-2.cs]: No idea. unload-2 does manage to print a fair number of dots before it dies, so maybe? :grin:

File: mono/tests/test-driver (hunk 2 above)
Comment (akoeplinger) [inline mono/tests/test-driver]: I think there were talks about moving them to the managed test runner, so it's good we're prepared :)

File: mono/tests/sgen-domain-unload-2.cs [hunk 8]
```diff
@@ -37,7 +40,33 @@ static void BackgroundNoise ()
 			AppDomain.Unload (ad);
 			Console.Write (".");
 			if (i > 0 && i % 20 == 0) Console.WriteLine ();
+			if (!testHasTimeLeft(startTime)) {
+				var finishTime = DateTime.UtcNow;
+				Console.WriteLine("stopping after {0}", finishTime-startTime);
```
Comment (akoeplinger) [inline mono/tests/sgen-domain-unload-2.cs]: would be nice to print `i` here as well so we get an idea of how many iterations it managed.

**Comment by kumpera** [comment]
Hey @akoeplinger,

I'm assigning you this PR, once you guys agree on its content, hit merge.

**Comment by lambdageek** [comment]
It's ready.  I want to see `sgen-new-threads-dont-join-stw.cs` print some progress on armel to make sure it's not stalled and then I'll do another PR to add the same timeout logic to it as `sgen-domain-unload-2.cs`

**Pull request merged**